Firestore/core/src/firebase/firestore/remote/online_state_tracker.h

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_ONLINE_STATE_TRACKER_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_REMOTE_ONLINE_STATE_TRACKER_H_
+
+#if !defined(__OBJC__)
+// TODO(varconst): make it a pure C++ file when it no longer needs to rely on
+// `FSTOnlineStateDelegate`.
+#error "This header only supports Objective-C++"
+#endif  // !defined(__OBJC__)
+
+#import <Foundation/Foundation.h>
+
+#include <string>
+
+#include "Firestore/core/src/firebase/firestore/model/types.h"
+#include "Firestore/core/src/firebase/firestore/util/async_queue.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+
+@protocol FSTOnlineStateDelegate;
+
+NS_ASSUME_NONNULL_BEGIN
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+/**
+ * A component used by the `FSTRemoteStore` to track the `OnlineState` (that is,
+ * whether or not the client as a whole should be considered to be online or
+ * offline), implementing the appropriate heuristics.
+ *
+ * In particular, when the client is trying to connect to the backend, we allow
+ * up to `kMaxWatchStreamFailures` within `kOnlineStateTimeout` for a connection
+ * to succeed. If we have too many failures or the timeout elapses, then we set
+ * the `OnlineState` to `Offline`, and the client will behave as if it is
+ * offline (`getDocument()` calls will return cached data, etc.).
+ */
+class OnlineStateTracker {
+ public:
+  OnlineStateTracker() = default;
+
+  explicit OnlineStateTracker(util::AsyncQueue* worker_queue,
+                              id<FSTOnlineStateDelegate> online_state_delegate)
```

Comment:
OnlineStateTracker on Android defines an embedded type specifically defined to allow the delegate to be passed as a lambda:

```java
  interface OnlineStateCallback {
    /**
     * Called whenever the online state of the client changes. This is based on the watch stream for
     * now.
     */
    void handleOnlineStateChange(OnlineState onlineState);
  }
```

Typescript takes this as an anonymous function type: `(onlineState: OnlineState) => void`.

I think we could reasonably make this take `std::function<void(OnlineState)>` and then supply an implementation that calls the delegate method in FSTRemoteStore. The tests can supply a no-op lambda rather than null.

Reply:
Done. I originally wanted to leave this off to the PRs dealing with `RemoteStore`, but it's true that it's not hard to implement now.